mage_ai/streaming/sources/nats_js.py

```diff
@@ -0,0 +1,187 @@
+            if self.config.nkeys_seed_str:
+                connect_opts["nkeys_seed_str"] = self.config.nkeys_seed_str
+
+            # Establish connection with the configured options
+            self.nc = await nats.connect(**connect_opts)
+            self.js = self.nc.jetstream()
+
+            # Default consumer_name to stream_name if not provided
+            consumer_name = self.config.consumer_name or self.config.stream_name
+
+            await self.js.add_stream(name=self.config.stream_name, subjects=[self.config.subject])
+            self.psub = await self.js.pull_subscribe(self.config.subject, consumer_name)
+        except NoServersError as e:
+            self._print(f'Caught NoServersError while connecting to NATS server: {e}')
+        except Exception as e:
+            self._print(f'Caught exception while connecting to NATS server: {e}')
+
+    # Define callback methods
+    async def disconnected_cb(self):
+        self._print('Got disconnected!')
+
+    async def reconnected_cb(self):
+        self._print(f'Got reconnected to {self.nc.connected_url.netloc}')
+
+    async def error_cb(self, e):
+        self._print(f'There was an error: {e}')
+
+    async def closed_cb(self):
+        self._print('Connection is closed')
+
+    async def aclose_client(self):
+        await self.nc.close()
+
+    def init_client(self):
+        future = asyncio.run_coroutine_threadsafe(self.ainit_client(), self.loop)
+        future.result()
+
+    def close_client(self):
+        asyncio.run_coroutine_threadsafe(self.aclose_client(), self.loop)
+
+    def batch_read(self, handler: Callable):
+        self.init_client()
+
+        try:
+            while True:
+                self._print("Fetching messages...")
+                message_tuples = self.fetch_messages()
+                self._print(f"Fetched {len(message_tuples)} messages")
+
+                if not message_tuples:
+                    self._print("No messages fetched, continuing to next iteration")
+                    continue
+
+                # Accumulate all messages in a batch
+                batch = [json.loads(msg.data.decode()) for _, msg in message_tuples]
+
+                self._print(f"Processing batch with {len(batch)} messages")
+
+                # Process the entire batch if it's not empty
+                if batch:
+                    try:
+                        handler(batch)
+                        self._print("Batch processed successfully, acknowledging messages")
+
+                        # Acknowledge all messages in the batch
+                        for _, msg in message_tuples:
+                            asyncio.run_coroutine_threadsafe(msg.ack(), self.loop)
+                    except Exception as e:
+                        self._print(f"Error processing batch: {e}")
+                        # Acknowledge only the messages processed successfully
+                        for _, msg in message_tuples[:len(batch)]:
```

Comment:
What's the difference between this and line 149? batch and message_tuples have same length

Reply:
addressed with 94d93afbd